prow/tide/tide.go

```diff
@@ -0,0 +1,289 @@
+			} else if overallState == successState && s == pendingState {
+				overallState = pendingState
+			}
+		}
+		c.log.Infof("PR #%d is in the %s state.", int(pr.Number), overallState)
+		if overallState == successState {
+			successes = append(successes, pr)
+		} else if overallState == pendingState {
+			pendings = append(pendings, pr)
+		} else {
+			nones = append(nones, pr)
+		}
+	}
+	if len(successes) > 0 {
+		c.log.Infof("Merge PR #%d.", int(pickSmallestNumber(successes).Number))
+	} else if len(pendings) > 0 {
+		c.log.Info("Do nothing. Waiting for pending PRs.")
+	} else if len(nones) > 0 {
+		c.log.Infof("Trigger tests for PR #%d.", int(pickSmallestNumber(nones).Number))
+	}
+	return nil
+}
+
+type subpool struct {
+	org    string
+	repo   string
+	branch string
+	sha    string
+	pjs    []kube.ProwJob
+	prs    []pullRequest
+}
+
+// dividePool splits up the list of pull requests and prow jobs into a group
+// per repo and branch. It only keeps ProwJobs that match the latest branch.
+func (c *Controller) dividePool(pool []pullRequest, pjs []kube.ProwJob) ([]subpool, error) {
+	sps := make(map[string]*subpool)
+	for _, pr := range pool {
+		org := string(pr.Repository.Owner.Login)
+		repo := string(pr.Repository.Name)
+		branch := string(pr.BaseRef.Name)
+		branchRef := string(pr.BaseRef.Prefix) + string(pr.BaseRef.Name)
+		fn := fmt.Sprintf("%s/%s %s", org, repo, branch)
+		if sps[fn] == nil {
+			sha, err := c.ghc.GetRef(org, repo, strings.TrimPrefix(branchRef, "refs/"))
+			if err != nil {
+				return nil, err
+			}
+			sps[fn] = &subpool{
+				org:    org,
+				repo:   repo,
+				branch: branch,
+				sha:    sha,
+			}
+		}
+		sps[fn].prs = append(sps[fn].prs, pr)
+	}
+	for _, pj := range pjs {
+		if pj.Spec.Type != kube.PresubmitJob && pj.Spec.Type != kube.BatchJob {
+			continue
+		}
+		fn := fmt.Sprintf("%s/%s %s", pj.Spec.Refs.Org, pj.Spec.Refs.Repo, pj.Spec.Refs.BaseRef)
+		if sps[fn] == nil || pj.Spec.Refs.BaseSHA != sps[fn].sha {
+			continue
+		}
+		sps[fn].pjs = append(sps[fn].pjs, pj)
+	}
+	var ret []subpool
+	for _, sp := range sps {
+		ret = append(ret, *sp)
+	}
+	return ret, nil
+}
+
+func (c *Controller) search(ctx context.Context, q string) ([]pullRequest, error) {
+	var ret []pullRequest
+	vars := map[string]interface{}{
+		"query":        githubql.String(q),
+		"searchCursor": (*githubql.String)(nil),
+	}
+	var totalCost int
+	var remaining int
+	for {
+		sq := searchQuery{}
+		if err := c.ghc.Query(ctx, &sq, vars); err != nil {
+			return nil, err
+		}
+		totalCost += int(sq.RateLimit.Cost)
+		remaining = int(sq.RateLimit.Remaining)
+		for _, n := range sq.Search.Nodes {
+			ret = append(ret, n.PullRequest)
+		}
+		if !sq.Search.PageInfo.HasNextPage {
+			break
+		}
+		vars["searchCursor"] = githubql.NewString(sq.Search.PageInfo.EndCursor)
+	}
+	c.log.Infof("Search for query \"%s\" cost %d point(s). %d remaining.", q, totalCost, remaining)
+	return ret, nil
+}
+
+type pullRequest struct {
```

Comment:
Nope, these aren't to be shared. The point of graphql is that each query asks for exactly the data that it needs.